### PR TITLE
Use tracing for format_dev

### PR DESCRIPTION
**Pull request by konstin**
## Summary

[tracing](https://github.com/tokio-rs/tracing) is library for logging, tracing and related features that has a large ecosystem. Using [tracing-subscriber](https://docs.rs/tracing-subscriber) and [tracing-indicatif](https://github.com/emersonford/tracing-indicatif), we get a nice logging output that you can configure with `RUST_LOG` (e.g. `RUST_LOG=debug`) and a live look into the formatter progress.

Default:
Screenshot from 2023-07-30 13-59-53

`RUST_LOG=debug`:
Screenshot from 2023-07-30 14-01-32

It's easy to see in this output which files take a disproportionate amount of time.

[Peek 2023-07-30 14-35.webm](https://github.com/astral-sh/ruff/assets/6826232/2c92db5c-1354-465b-a6bc-ddfb281d6f9d)

It opens up further integration with the tracing ecosystem, [tracing-timing](https://docs.rs/tracing-timing/latest/tracing_timing/) and [tokio-console](https://github.com/tokio-rs/console) can e.g. show histograms and the json output allows us building better pipelines than grepping a log file.

One caveat is using `parent: None` for the logging statements because tracing subscriber does not allow deactivating the span without reimplementing all the other log message formatting, too, and we don't need span information, esp. since it would currently show the progress bar span.

## Test Plan

n/a

### CODE REVIEW

File: crates/ruff_dev/src/format_dev.rs [hunk 1]
```diff
@@ -239,78 +261,88 @@ fn format_dev_multi_project(args: &Args) -> anyhow::Result<bool> {
...
+                info!(
+                    parent: None,
+                    file_count = result.file_count,
+                    similarity_index = result.statistics.similarity_index(),
+                    duration = result.duration.as_secs_f32(),
+                    "Finished {}: {} files, similarity index {:.3}, {:.2}s",
+                    project_path.display(),
+                    result.file_count,
+                    result.statistics.similarity_index(),
+                    result.duration.as_secs_f32(),
```
Comment (konstin) [inline crates/ruff_dev/src/format_dev.rs:281]: was experimenting with tracing here

File: crates/ruff_dev/src/format_dev.rs [hunk 2]
```diff
@@ -239,78 +261,88 @@ fn format_dev_multi_project(args: &Args) -> anyhow::Result<bool> {
         None => None,
     };
 
-    #[allow(clippy::print_stdout)]
     for project_path in project_paths {
-        bar.suspend(|| println!("Starting {}", project_path.display()));
```
Comment (konstin) [inline crates/ruff_dev/src/format_dev.rs:244]: a nice side effect is that tracing-indicatif makes sure the progress bars and the logging output don't collide

File: crates/ruff_dev/src/format_dev.rs [hunk 3]
```diff
@@ -239,78 +261,88 @@ fn format_dev_multi_project(args: &Args) -> anyhow::Result<bool> {
...
         }
-        bar.finish_and_clear();
     }
 
+    drop(pb_span_enter);
+    drop(pb_span);
+
     let duration = start.elapsed();
 
-    #[allow(clippy::print_stdout)]
```
Comment (konstin) [inline crates/ruff_dev/src/format_dev.rs:283]: another positive side effect: no more manual stdout/stdout

File: crates/ruff_dev/Cargo.toml [hunk 4]
```diff
@@ -42,6 +42,9 @@ strum = { workspace = true }
 strum_macros = { workspace = true }
 tempfile = "3.6.0"
 toml = { workspace = true, features = ["parse"] }
+tracing = { workspace = true }
```
Comment (MichaReiser) [inline crates/ruff_dev/Cargo.toml:45]: Can we remove `log` or is it used by other dev tools?

File: crates/ruff_dev/src/format_dev.rs [hunk 5]
```diff
@@ -181,23 +186,38 @@ pub(crate) struct Args {
...
+        .unwrap();
+    tracing_subscriber::registry()
+        .with(filter_layer)
+        .with(
+            tracing_subscriber::fmt::layer()
+                .with_writer(indicatif_layer.get_stderr_writer())
+                .with_target(false),
+        )
+        .with(indicatif_layer)
+        .init();
```
Comment (MichaReiser) [inline crates/ruff_dev/src/format_dev.rs:202]: Nit: Maybe move out to its own function to not clutter `main` as much.

File: crates/ruff_dev/src/format_dev.rs [hunk 6]
```diff
@@ -239,78 +261,88 @@ fn format_dev_multi_project(args: &Args) -> anyhow::Result<bool> {
...
+    );
 
     // TODO(konstin): black excludes
 
     // Find files to check (or in this case, format twice). Adapted from ruff_cli
     // First argument is ignored
     let paths = ruff_check_paths(files)?;
 
-    let bar = progress_bar.then(|| ProgressBar::new(paths.len() as u64));
+    let pb_span = info_span!("format_dev_project progress bar", first_file = %files[0].display());
```
Comment (MichaReiser) [inline crates/ruff_dev/src/format_dev.rs:341]: Nit: You could wrap line 341..398 in a block, this way you don't need the explicit drop.

File: crates/ruff_dev/src/format_dev.rs [hunk 7]
```diff
@@ -239,78 +261,88 @@ fn format_dev_multi_project(args: &Args) -> anyhow::Result<bool> {
...
-                        "Finished {}: {} files, similarity index {:.3}, {:.2}s",
-                        project_path.display(),
-                        result.file_count,
-                        result.statistics.similarity_index(),
-                        result.duration.as_secs_f32(),
+                info!(
+                    parent: None,
+                    file_count = result.file_count,
+                    similarity_index = result.statistics.similarity_index(),
+                    duration = result.duration.as_secs_f32(),
```
Comment (MichaReiser) [inline crates/ruff_dev/src/format_dev.rs:276]: Why are we passing all these arguments here? Do we want to filter by these?

**Review by MichaReiser — APPROVED** [review]
Love it. We should use tracing in all of ruff ;)

We used the tracing timing in rome to understand formatting/linting times better (down to per rule). That was very useful

File: crates/ruff_dev/Cargo.toml (hunk 4 above)
Comment (konstin) [reply crates/ruff_dev/Cargo.toml:45]: yes!

File: crates/ruff_dev/src/format_dev.rs (hunk 7 above)
Comment (konstin) [reply crates/ruff_dev/src/format_dev.rs:276]: i was experimenting with progress bars here

**Comment by konstin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6177** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6177" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/6177?utm_source=stack-comment).

**Pull request merged**